src/shared/components/data-table/components/table-controls/components/table-actions/components/download-data/DownloadData.tsx

```diff
@@ -0,0 +1,134 @@
+ * limitations under the License.
+ */
+import React, { ReactNode, useEffect, useRef } from 'react';
+import ReactDOM from 'react-dom/client';
+import memoize from 'lodash/memoize';
+
+import useDataTable from 'src/shared/components/data-table/hooks/useDataTable';
+import { downloadTextAsFile } from 'src/shared/helpers/downloadAsFile';
+
+import LoadingButton from 'src/shared/components/loading-button';
+
+import { TableAction } from 'src/shared/components/data-table/dataTableTypes';
+
+import styles from './DownloadData.scss';
+
+const getReactRenderer = memoize(() => {
+  const element = document.createElement('div');
+  const root = ReactDOM.createRoot(element);
+
+  return {
+    element,
+    renderer: root
+  };
+});
+
+const getReactNodeText = (node: ReactNode): string => {
+  const { element, renderer } = getReactRenderer();
+  renderer.render(node);
+  return element.innerText;
+};
+
+const DownloadData = () => {
+  const {
+    dispatch,
+    rows,
+    downloadFileName,
+    columns,
+    selectedAction,
+    hiddenRowIds,
+    downloadHandler
+  } = useDataTable();
+
+  const allowComponentResetRef = useRef(true);
+
+  useEffect(() => {
+    return () => {
+      allowComponentResetRef.current = false;
+    };
+  }, []);
+
+  const restoreDefaults = () => {
+    setTimeout(() => {
+      if (allowComponentResetRef.current) {
+        dispatch({
+          type: 'set_selected_action',
+          payload: TableAction.DEFAULT
+        });
+      }
+    }, 1000);
+  };
+
+  const handleDownload = async () => {
+    if (downloadHandler) {
+      await downloadHandler();
+      setTimeout(restoreDefaults, 1000);
+      return;
+    }
+    const dataForExport: string[][] = [];
+    dataForExport[0] = [
+      ...columns
+        .filter((column) => column.isExportable !== false)
+        .map((column) => column.title ?? '')
+    ];
+
+    const rowsToDownload =
+      selectedAction === TableAction.DOWNLOAD_ALL_DATA
+        ? rows
+        : rows.filter((row) => !hiddenRowIds[row.rowId]);
+
+    rowsToDownload.forEach((row, rowIndex) => {
+      dataForExport[rowIndex + 1] = [];
+      row.cells.forEach((cell, cellIndex) => {
+        const { renderer, isExportable } = columns[cellIndex];
+
+        if (isExportable !== false) {
+          const cellExportData = renderer
+            ? renderer({
+                rowData: row.cells,
+                rowId: String(row.rowId),
+                cellData: cell
+              })
+            : cell;
+
+          dataForExport[rowIndex + 1].push(getReactNodeText(cellExportData));
+        }
+      });
+    });
+
+    const csv = formatCSV(dataForExport);
+
+    downloadTextAsFile(csv, downloadFileName ?? 'Table export.csv');
+
+    restoreDefaults();
+  };
+
+  return (
+    <div className={styles.downloadData}>
+      <span>{downloadFileName ?? 'table.csv'}</span>
+      <LoadingButton onClick={handleDownload}>Download</LoadingButton>
```

Comment:
This is super minor, but since all data is already on the client, the download completes before the spinner stops:

https://user-images.githubusercontent.com/6834224/195536179-bd8beb17-a4f6-48d5-af6d-76a993db3b6e.mov

It can be addressed by changing from LoadingButton to ControlledLoadingButton, and driving the state of that button from the `DownloadData` component. Or we can ignore this for now, and see how https://github.com/Ensembl/ensembl-client/pull/841 pans out.

Reply:
I have addressed this issue in the latest commit.